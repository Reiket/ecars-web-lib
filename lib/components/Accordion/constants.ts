export const ACCORDION_TEST_ID = 'accordionTestId';
