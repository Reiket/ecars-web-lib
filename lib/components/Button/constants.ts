import type {ButtonHTMLAttributes} from 'react';

export const BUTTON_SIZE = Object.freeze({
  BIG: 'big',
  SMALL: 'small',
});

export const BUTTON_COLOR = Object.freeze({
  GREEN: 'green',
  WHITE: 'white',
  GRAY: 'gray',
});


export type ButtonAttributesType = ButtonHTMLAttributes<HTMLButtonElement>['type']
export type ButtonSizeType = (typeof BUTTON_SIZE)[keyof typeof BUTTON_SIZE];
export type ButtonColorType = (typeof BUTTON_COLOR)[keyof typeof BUTTON_COLOR];

export const BUTTON_TEST_ID = 'buttonTestId';