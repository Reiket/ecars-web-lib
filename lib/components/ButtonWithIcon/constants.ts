export const LEFT_ICON_BUTTON_TEST_ID = 'leftIconButtonTestId';
export const RIGHT_ICON_BUTTON_TEST_ID = 'rightIconButtonTestId';
