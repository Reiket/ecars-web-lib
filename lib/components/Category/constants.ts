export const CATEGORIES_LIST = Object.freeze({
  GUIDES: 'guides',
  REVIEWS: 'reviews',
  NEWS: 'news',
});

export type CategoriesListType = (typeof CATEGORIES_LIST)[keyof typeof CATEGORIES_LIST];

export const CATEGORY_TEST_ID = 'categoryTestId';
