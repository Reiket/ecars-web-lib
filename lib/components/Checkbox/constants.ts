import type {ElementProps} from '@/services/types';

export interface CheckboxProps extends ElementProps {
  name?: string;
  id?: string;
  disabled?: boolean;
  onChange?: (checked: boolean) => void;
  checked?: boolean;
  hasError?: boolean;
}

export const CHECKBOX_TEST_ID = 'checkboxTestId';
export const CHECKBOX_TEST_INPUT_ID = 'checkboxTestInputId';
