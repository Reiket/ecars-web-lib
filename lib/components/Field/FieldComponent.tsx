import type {ReactElement, FC} from 'react';
import {cloneElement} from 'react';
import {Field, FIELD_TEST_ID} from '@/components/Field/constants';
import type {ElementProps} from '@/services/types';
import {cn} from '@/services/helpers';

export interface Props extends ElementProps {
  children: ReactElement;
  id?: string;
  label?: string;
  error?: string;
}

export const FieldComponent: FC<Props> = ({label, block, children, error, id = 'input'}) => {
  const classNames = cn(block, 'field');
  return (
    <div
      data-testid={FIELD_TEST_ID}
      className={classNames}
    >
      {!!label && (
        <Field.Label
          text={label}
          htmlFor={id}
        />
      )}
      {cloneElement(children, {id, hasError: !!error})}
      {!!error && <Field.Error errorText={error} />}
    </div>
  );
};
