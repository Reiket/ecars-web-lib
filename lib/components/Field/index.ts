import {FieldComponent} from '@/components/Field/FieldComponent';
import {FieldLabel} from '@/components/Field/components/FieldLabel';
import {FieldError} from '@/components/Field/components/FieldError';

export const Field = Object.assign(FieldComponent, {
  Label: FieldLabel,
  Error: FieldError,
});
