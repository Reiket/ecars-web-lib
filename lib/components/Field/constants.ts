import {FieldComponent} from '@/components/Field/FieldComponent';
import {FieldError} from '@/components/Field/components/FieldError';
import {FieldLabel} from '@/components/Field/components/FieldLabel';

export const Field = Object.assign(FieldComponent, {
  Label: FieldLabel,
  Error: FieldError,
});

export const FIELD_TEST_ID = 'fieldTestId';
export const FIELD_LABEL_TEST_ID = 'fieldLabelTestId';
