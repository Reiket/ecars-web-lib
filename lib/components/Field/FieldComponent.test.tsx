import {render, screen} from '@testing-library/react';
import {Input} from '@/components/Input/Input';
import {Field, FIELD_LABEL_TEST_ID, FIELD_TEST_ID} from '@/components/Field/constants';
import {INPUT_TEST_ID} from '@/components/Input/constants';
import '@testing-library/jest-dom';

describe('Field Component', () => {
  const labelOptions = ['Email', undefined];
  const errorOptions = ['Some error', undefined];
  const idOptions = ['email', undefined];
  labelOptions.forEach((label) => {
    errorOptions.forEach((error) => {
      idOptions.forEach((id) => {
        const errorStatus = error ? `with error - ${error}` : 'without error';
        const labelStatus = label ? `with label - ${label}` : 'without label';
        const idStatus = id ? `with id - ${id}` : 'without id';
        const testName = `render field component ${errorStatus}, ${labelStatus}, ${idStatus}`;
        test(testName, () => {
          const {container} = render(
            <Field
              label={label}
              id={id}
              error={error}
            >
              <Input />
            </Field>,
          );
          const fieldElement = screen.getByTestId(FIELD_TEST_ID);
          const fieldLabelElement = screen.queryByTestId(FIELD_LABEL_TEST_ID);
          const inputElement = screen.getByTestId(INPUT_TEST_ID);
          expect(container).toMatchSnapshot();
          expect(fieldElement).toBeInTheDocument();
          expect(inputElement).toBeInTheDocument();
          if (label) {
            expect(fieldLabelElement).toBeInTheDocument();
          }
          if (id) {
            expect(fieldLabelElement).toHaveAttribute('htmlFor', id);
            expect(inputElement).toHaveAttribute('id', id);
          }
          if (error) {
            expect(screen.getByText('Some error')).toBeInTheDocument();
            expect(inputElement).toHaveClass('error');
          }
        });
      });
    });
  });
});
