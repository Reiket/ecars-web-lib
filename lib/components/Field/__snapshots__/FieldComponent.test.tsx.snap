// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Component render field component with error - Some error, with label - Email, with id - email 1`] = `
<div>
  <div
    class="field"
    data-testid="fieldTestId"
  >
    <label
      class="field__label"
      data-testid="fieldLabelTestId"
      for="email"
    >
      Email
    </label>
    <input
      autocomplete="off"
      class="input error"
      data-testid="inputTestId"
      id="email"
      name="input"
      type="text"
      value=""
    />
    <p
      class="field__error"
    >
      Some error
    </p>
  </div>
</div>
`;

exports[`Field Component render field component with error - Some error, with label - Email, without id 1`] = `
<div>
  <div
    class="field"
    data-testid="fieldTestId"
  >
    <label
      class="field__label"
      data-testid="fieldLabelTestId"
      for="input"
    >
      Email
    </label>
    <input
      autocomplete="off"
      class="input error"
      data-testid="inputTestId"
      id="input"
      name="input"
      type="text"
      value=""
    />
    <p
      class="field__error"
    >
      Some error
    </p>
  </div>
</div>
`;

exports[`Field Component render field component with error - Some error, without label, with id - email 1`] = `
<div>
  <div
    class="field"
    data-testid="fieldTestId"
  >
    <input
      autocomplete="off"
      class="input error"
      data-testid="inputTestId"
      id="email"
      name="input"
      type="text"
      value=""
    />
    <p
      class="field__error"
    >
      Some error
    </p>
  </div>
</div>
`;

exports[`Field Component render field component with error - Some error, without label, without id 1`] = `
<div>
  <div
    class="field"
    data-testid="fieldTestId"
  >
    <input
      autocomplete="off"
      class="input error"
      data-testid="inputTestId"
      id="input"
      name="input"
      type="text"
      value=""
    />
    <p
      class="field__error"
    >
      Some error
    </p>
  </div>
</div>
`;

exports[`Field Component render field component without error, with label - Email, with id - email 1`] = `
<div>
  <div
    class="field"
    data-testid="fieldTestId"
  >
    <label
      class="field__label"
      data-testid="fieldLabelTestId"
      for="email"
    >
      Email
    </label>
    <input
      autocomplete="off"
      class="input"
      data-testid="inputTestId"
      id="email"
      name="input"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`Field Component render field component without error, with label - Email, without id 1`] = `
<div>
  <div
    class="field"
    data-testid="fieldTestId"
  >
    <label
      class="field__label"
      data-testid="fieldLabelTestId"
      for="input"
    >
      Email
    </label>
    <input
      autocomplete="off"
      class="input"
      data-testid="inputTestId"
      id="input"
      name="input"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`Field Component render field component without error, without label, with id - email 1`] = `
<div>
  <div
    class="field"
    data-testid="fieldTestId"
  >
    <input
      autocomplete="off"
      class="input"
      data-testid="inputTestId"
      id="email"
      name="input"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`Field Component render field component without error, without label, without id 1`] = `
<div>
  <div
    class="field"
    data-testid="fieldTestId"
  >
    <input
      autocomplete="off"
      class="input"
      data-testid="inputTestId"
      id="input"
      name="input"
      type="text"
      value=""
    />
  </div>
</div>
`;
